### PR TITLE
Pluggable rescorers

### PR DESCRIPTION
**Pull request by anti-social**
Prerequisites: In our company we need specific scoring. We have a lot of suppliers with huge amount of goods. Some suppliers are "better" than others. But if supplier rank affects score, goods from top ranked supplier supersede goods from other suppliers. What we really want to get is:
```
product1 (supplier1)
product2 (supplier2)
product3 (supplier3)
product4 (supplier1)
product5 (supplier2)
...
```

The most suitable and simple way to do that is to write specific rescorer (the only requirement that all products from specific supplier should be located on the same shard). Thus we decided to make rescorers pluggable. There were an [issue](#17331) and a [discussion](https://discuss.elastic.co/t/how-to-plug-in-alternative-rescorer/11555) for this feature.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/search/rescore/RescorePluginIT.java [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+                        "post",
+                        jsonBuilder()
+                        .startObject().startObject("post").startObject("properties")
+                        .startObject("title")
+                                .field("type", "text")
+                        .endObject()
+                        .endObject().endObject().endObject())
+                .execute()
+                .actionGet();
+        client().admin().cluster()
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/rescore/RescorePluginIT.java]: This shouldn't be needed anymore. By default we wait for the index to be created now.

File: core/src/test/java/org/elasticsearch/search/rescore/RescorePluginIT.java [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+                .setSettings(SETTING_NUMBER_OF_SHARDS, 1)
+                .addMapping(
+                        "post",
+                        jsonBuilder()
+                        .startObject().startObject("post").startObject("properties")
+                        .startObject("title")
+                                .field("type", "text")
+                        .endObject()
+                        .endObject().endObject().endObject())
+                .execute()
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/rescore/RescorePluginIT.java]: `.get()` is more idiomatic now.

File: core/src/test/java/org/elasticsearch/search/rescore/RescorePluginIT.java [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+                .actionGet();
+
+        client().index(indexRequest("test").type("post").id("1").source(
+                        jsonBuilder()
+                        .startObject()
+                                .field("title", "The quick brown")
+                        .endObject()))
+                .actionGet();
+        client().index(indexRequest("test").type("post").id("2").source(
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/rescore/RescorePluginIT.java]: `client().prepareIndex(...` is more normal now.

**Review by nik9000 — CHANGES_REQUESTED** [review]
I left a few minor things but this looks just about right.

I'm sorry this took so long to review.

File: core/src/main/java/org/elasticsearch/search/rescore/RescoreParser.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.index.query.QueryParseContext;
+
+import java.io.IOException;
+
+/**
+ * Parses XContent into a {@link RescoreBuilder}.
+ */
+@FunctionalInterface
+public interface RescoreParser<RB extends RescoreBuilder<RB>> {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/rescore/RescoreParser.java]: Note to self: recheck if we need this. I *think* we can reuse some interface but I don't know off hand.

**Comment by jimczi** [comment]
> The most suitable and simple way to do that is to write specific rescorer (the only requirement that all products from specific supplier should be located on the same shard). Thus we decided to make rescorers pluggable.

I don't think it's the most suitable way. Rescorer works only on a per shard basis with a fixed window size. What happens if you have a single supplier that fills the window size entirely on a single shard ? How do you set the window size ?
The fact that rescorers are not pluggable is well explained in the discussion that you linked. Rescoring with a query is "ok" because it works like the normal top docs and does not rely on how we merge the documents. IMO for this to work you should not require routing or setting a window size. In fact your problem could fit more nicely within a `TopDocs` implementation like the `collapse` feature does:
https://www.elastic.co/guide/en/elasticsearch/reference/5.3/search-request-collapse.html
@markharwood worked on some custom top docs collector for aggregations too. Maybe we can reuse some of them here:
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-diversified-sampler-aggregation.html

IMO this is too dangerous to let users implement rescorer freely especially if they need to know a lot of internals to do it.  The majority of use cases are already covered with query rescorers and for this specific problem I don't think rescorer are a good fit.

Though I do agree with @nik9000 that the PR looks great if we'd have to add this feature, it's just that I don't like what it allows ;)

**Comment by rjernst** [comment]
I've marked this for discussion as I think a decision should be made on whether this is a direction we want to go.

**Comment by rjernst** [comment]
@jimczi @nik9000 @anti-social I think this was superceded by #26368?

**Pull request closed without merging**